include/zephyr/bluetooth/gatt.h

```diff
@@ -165,21 +165,117 @@ typedef ssize_t (*bt_gatt_attr_write_func_t)(struct bt_conn *conn,
 					     const void *buf, uint16_t len,
 					     uint16_t offset, uint8_t flags);
 
-/** @brief GATT Attribute structure. */
+/** @brief GATT Attribute
+ *
+ *  This type primarily represents an ATT Attribute that may be
+ *  an entry in the local ATT database. The objects of this type
+ *  must be part of an array that forms a GATT service.
+ *
+ *  While the formed GATT service is registered with the local
+ *  GATT server, pointers to this type can typically be given to
+ *  GATT server APIs, like bt_gatt_notify().
+ *
+ *  @note This type is given as an argument to the
+ *  bt_gatt_discover() application callback, but it's not a
+ *  proper object of this type. The permissions, read() and
+ *  write() methods are not available, and it's unsound to pass
+ *  the pointer to GATT server APIs.
+ */
 struct bt_gatt_attr {
-	/** Attribute UUID */
+	/** @brief Attribute Type, aka. "UUID"
+	 *
+	 *  The Attribute Type determines the interface can be expected
+	 *  from the read() and write() methods and the possible
+	 *  permission configurations.
+	 *
+	 *  E.g. Attribute of type @ref BT_UUID_GATT_CPF will act as a
+	 *  GATT Characteristic Presentation Format descriptor as
+	 *  specified in Core Specification 3.G.3.3.3.5.
+	 *
+	 *  Anyone can define a new Attribute Type.
+	 */
 	const struct bt_uuid *uuid;
-	/** Attribute read callback */
+
+	/** @brief Attribute Value read method
```

Comment:
We could consider also documenting that the return value from this callback function is sent to the remote device (if any) as a read response. 

Similarly we could document that calling e.g. `bt_gatt_notify` in this callback will send the notification before the read response, and possible an example on how to postpone the notification until after the read response (as that is a requirement for several service specifications).

Reply:
Should we mention `bt_gatt_attr_read` as a helper function to deal with reads?

Reply:
> We could consider also documenting that the return value from this callback function is sent to the remote device (if any) as a read response.

We do, I think, a few lines below:

> This function is primarily invoked by the GATT server when a remote GATT client performs a read

Is this understandable enough?

Reply:
> Similarly we could document that calling e.g. `bt_gatt_notify` in this callback will send the notification before the read response, and possible an example on how to postpone the notification until after the read response (as that is a requirement for several service specifications).

This is too complicated to add in this PR. Can we discuss later it and follow up?

Reply:
> Is this understandable enough?

It does not mention that the return value (which is only defined in the typedef) is sent over air to the client. I think it's valuable to add that so that users don't think that the return value is only a local value.

Reply:
> It does not mention that the return value (which is only defined in the typedef) is sent over air to the client. I think it's valuable to add that so that users don't think that the return value is only a local value.

I added the following text now.

> The GATT server propagates the return value from this method back to the remote client.